op-node/rollup/derive/span_batch.go

```
@@ -493,59 +500,88 @@ func (b *SpanBatch) GetBlockCount() int {
+	// always update the caches
+	defer b.updateCaches()
+
+	// if there is only one batch, initialize the originChangedBit and ParentCheck
+	if len(b.Batches) == 1 {
+		var originChanged uint
+		if seqNum == 0 {
+			originChanged = 1
+		}
+		b.SetFirstOriginChangedBit(originChanged)
+		copy(b.ParentCheck[:], singularBatch.ParentHash.Bytes()[:20])
+		return
+	}
+	// if the last two blocks are not ordered, the structure is not ordered
+	if b.Batches[len(b.Batches)-2].Timestamp > b.Batches[len(b.Batches)-1].Timestamp {
+		// TODO: this could be recovered by reordering the blocks, and rebuilding the caches
+		panic("span batch is not ordered")
+	}
+}
+
+// UpdateCaches updates caches based on the last batch added
+func (b *SpanBatch) updateCaches() {
+	// set the originBit for the new batch according to if the epoch number has changed
+	epochBit := uint(0)
+	// if there is only one batch, it is already managed by SetFirstOriginChangedBit
+	if len(b.Batches) > 1 {
+		if b.Batches[len(b.Batches)-2].EpochNum < b.Batches[len(b.Batches)-1].EpochNum {
+			epochBit = 1
+		}
+		b.originBits.SetBit(&b.originBits, len(b.Batches)-1, epochBit)
+	}
+
+	// update the blockTxCounts cache with the latest batch's tx count
+	b.blockTxCounts = append(b.blockTxCounts, uint64(len(b.Batches[len(b.Batches)-1].Transactions)))
+
+	// add each tx to the txs cache
+	// and to the spanBatchTxs cache
+	newTxs := [][]byte{}
+	for i := 0; i < len(b.Batches[len(b.Batches)-1].Transactions); i++ {
+		newTxs = append(newTxs, b.Batches[len(b.Batches)-1].Transactions[i])
+	}
+	// if the txs fail to add to the cache, store the error
+	// it can be returned when ToRawSpanBatch is called
+	// TODO: this is not ideal. If callers update caches repeatedly, they may waste time before finding out the error.
+	// in practice, ToRawSpanBatch is called after every AppendSingularBatch, so it should be fine.
+	if err := b.sbtxs.AddTxs(newTxs, b.ChainID); err != nil {
+		b.storedSbtxErr = err
+	}
```

IMO we should just bubble up the error to the caller and add an error to `updateChaches` and `AppendSingularBatch`. The critical caller is `SpanChannelOut.AddSingularBatch` and this function already returns an error. Would also remove the code smell to store the error in an error variable.